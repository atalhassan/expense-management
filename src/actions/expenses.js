import uuid from 'uuid'
import database from '../firebase/firebase'

// ADD_EXPENSES
export const addExpenses = (expense) => ({
  type: 'ADD_EXPENSES',
  expense
})
// Starting the add process
export const startAddExpense = (expenseData) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    const {
      description= '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = {description, note, amount, createdAt}
    return database.ref(`users/${uid}/expenses`).push(expense).then((snapshot) => {
      dispatch(addExpenses({
        id: snapshot.key,
        ...expense
      }))
    })
  }
}

// REMOVE_EXPENSES
export const removeExpenses = ({id}) => ({
  type: 'REMOVE_EXPENSES',
  id
})

// Starting the remove process
export const startRemoveExpense = ({id}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/expenses/${id}`).remove().then(() => {
      dispatch(removeExpenses({id}))
    })
  }
}

// EDIT_EXPENSES
export const editExpenses = (id, updates) => ({
  type: 'EDIT_EXPENSES',
  id,
  updates
})

// Starting the edit process
export const startEditExpense = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/expenses/${id}`).update(updates).then(() => {
      dispatch(editExpenses(id, updates))
    })
  }
}

// SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
})

export const startSetExpenses = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/expenses`)
      .once('value')
      .then((snapshot) => {
        const expenses = []
        snapshot.forEach((childSnapshot) => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          })
        })
        dispatch(setExpenses(expenses))
      })
  }
}
