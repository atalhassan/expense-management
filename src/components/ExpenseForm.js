import React from 'react'
import moment from 'moment'
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';


class ExpenseForm extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      description: props.expense ? props.expense.description : '',
      amount: props.expense ? (props.expense.amount /100).toString() : '',
      note: props.expense ? props.expense.note : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ''
    }
  }
  changeByName = (e) => {
    const change = {}
    const name = e.target.name
    change[name] = e.target.value
    if (name === 'amount') {
      const amount = e.target.value
      if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
        change[name] = amount
      } else {
        change[name] = this.state.amount
      }
      this.setState(() => change)
    } else {
      this.setState(() => change)
    }
  }
  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({createdAt}))
    }
  }
  onFocusChange = ({focused}) => {
    this.setState(() => ({calendarFocused: focused}))
  }
  onSubmit = (e) => {
    e.preventDefault()
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({error: 'please fill the description and amount'}))
    } else {
      this.setState(() => ({error: ''}))
      this.props.onSubmit({description: this.state.description,
        amount:parseFloat(this.state.amount, 10) * 100,
        createdAt:this.state.createdAt.valueOf(),
        note: this.state.note})
      }
  }

  render() {
    return (
      <form className='form' onSubmit={this.onSubmit}>
        {this.state.error && <p className='form__error'>{this.state.error}</p>}
        <input
          type='text'
          name='description'
          placeholder='Description'
          value={this.state.description}
          onChange={this.changeByName}
          className='text-input'
          autoFocus
        />
        <input
          type='text'
          name='amount'
          placeholder='amount'
          value={this.state.amount}
          className='text-input'
          onChange={this.changeByName}
        />
        <SingleDatePicker
          date={this.state.createdAt}
          onDateChange={this.onDateChange}
          focused={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        <textarea
          placeholder='Add a note for your expense (optional)'
          name='note'
          className='text-area '
          value={this.state.note}
          onChange={this.changeByName}
        >
        </textarea>
        <div>
          {this.props.fromEdit ? <button className='button'>Save Expense</button> : <button className='button'>Add Expense</button>}
        </div>
      </form>
    )
  }

}


export default ExpenseForm
