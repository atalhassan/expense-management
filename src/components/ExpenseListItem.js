import React from 'react'
import {Link} from 'react-router-dom'
import moment from 'moment'
import {connect} from 'react-redux'

export const ExpenseListItem = ({id, description, amount, createdAt}) => (
  <div>
      <Link to={`/edit/${id}`}>
        <h3>{description}</h3>
      </Link>
      <p>${parseFloat(amount/100, 10)} - {moment(createdAt).format('MMM Do, YYYY')}</p>

  </div>
);



export default connect()(ExpenseListItem)
