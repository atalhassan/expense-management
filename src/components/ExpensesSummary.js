import React from 'react'
import numeral from 'numeral'
import {connect} from 'react-redux'
import {Link} from 'react-router-dom'
import getVisibleExpenses from '../selectors/expenses'
import getVisibleExpensesTotal from '../selectors/expense-total'

export const ExpensesSummary = ({expenseCount, expenseTotal}) => {
  const expenseWord = expenseCount === 1 ? 'expense' : 'expenses'
  const formattedExpenseTotal = numeral(expenseTotal/100).format('$0,0.00')
  return (
    <div className='page-header'>
      <div className='content-container'>
        <h1 className='page-header__title'>Viewing <span>{expenseCount}</span> {expenseWord} totalling <span>{formattedExpenseTotal}</span></h1>
        <div className='page-header__action'>
          <Link className='button' to='/create'>Add Expense</Link>
        </div>
      </div>
    </div>
  )
}

const mapStateToProps = (state) => {
  const visibleExpenses = getVisibleExpenses(state.expenses, state.filters)
  return {
    expenseCount: visibleExpenses.length,
    expenseTotal: getVisibleExpensesTotal(visibleExpenses)
  }
}

export default connect(mapStateToProps)(ExpensesSummary)
