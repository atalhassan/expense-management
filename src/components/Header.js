import React from 'react';
import {NavLink} from 'react-router-dom'

const Header = () => (
  <header>
    <h1>Expensify</h1>
    <NavLink to='/' activeClassName='is-action' exact>Dashboard </NavLink>
    <NavLink to='/create' activeClassName='is-action' >Create </NavLink>

  </header>
)

export default Header;
