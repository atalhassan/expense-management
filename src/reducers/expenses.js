
const expensesReducerDefaultState = []

const expensesReducer = (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSES':
      return [...state, action.expense]
    case 'REMOVE_EXPENSES':
      return state.filter(({id}) => id !== action.id)
    case 'EDIT_EXPENSES':
      return state.map((expense) => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.updates
          }
        } else {
          return expense
        }
      })
    default:
      return state
  }
}

export default expensesReducer;
