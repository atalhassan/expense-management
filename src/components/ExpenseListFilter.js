import React from 'react'
import {connect} from 'react-redux'
import {setTextFilter, SetEndDate, SetStartDate, SortByAmount, SortByDate} from '../actions/filters'

// import 'react-dates/initialize';
import { DateRangePicker } from 'react-dates';
// import 'react-dates/lib/css/_datepicker.css';

export class ExpenseListFilter extends React.Component {
  state = {
    calendarFocused: null
  }
  onDatesChange = ({startDate, endDate}) => {
    this.props.setStartDate(startDate)
    this.props.setEndDate(endDate)
  }
  onFocusChange = (calendarFocused) => {
    this.setState(() => ({calendarFocused}))
  }
  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value)
  }
  onSortChange = (e) => {
    if (e.target.value === 'date') {
      this.props.SortByDate()
    } else if (e.target.value === 'amount') {
      this.props.SortByAmount()
    }
  }
  render() {
    return (
      <div className='content-container'>
        <div className='input_group'>
          <div className='input_group__item'>
            <input
              type='text'
              className='text-input'
              value={this.props.filters.text}
              placeholder='Search expenses'
              onChange={this.onTextChange}
            />

          </div>
          <div className='input_group__item'>
            <select
              className='select'
              value={this.props.filters.sortBy}
              onChange={this.onSortChange}
              >
                <option value='date'>Date</option>
                <option value='amount'>Amount</option>
              </select>
            </div>
            <div className='input_group__item'>
              <DateRangePicker
                startDate={this.props.filters.startDate}
                endDate={this.props.filters.endDate}
                onDatesChange={this.onDatesChange}
                focusedInput={this.state.calendarFocused}
                onFocusChange={this.onFocusChange}
                showClearDates={true}
                numberOfMonths={1}
                isOutsideRange={() => false}
              />
            </div>
          </div>
        </div>
      )
    }
  }


  const mapStateToProps = (state) => {
    return {
      filters: state.filters
    }
  }

  const mapDispatchToProps = (dispatch) => {
    return {
      setTextFilter : (text) => dispatch(setTextFilter(text)),
      SortByDate : () => dispatch(SortByDate()),
      SortByAmount: () => dispatch(SortByAmount()),
      setStartDate: (startDate) => dispatch(SetStartDate(startDate)),
      setEndDate: (endDate) => dispatch(SetEndDate(endDate))
    }
  }

  export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilter);
