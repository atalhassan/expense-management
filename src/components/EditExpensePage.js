import React from 'react';
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm'
import {removeExpenses, editExpenses} from '../actions/expenses'

export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.editExpenses(this.props.expense.id,{...expense})
    this.props.history.push('/')
  }
  onClick = () => {
    this.props.removeExpenses({id: this.props.expense.id})
    this.props.history.push('/')
  }

  render()  {
    return (
      <div>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onSubmit}
        />
        <button
          onClick={this.onClick}
          >Remove</button>
      </div>
    )
  }
}

const mapStateToProp = (state, props) => {
  return {
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  }
}

const mapDispatchToProps = (dispatch, props) => {
  return {
    editExpenses: (expense) => dispatch(editExpenses(props.expense.id,{...expense})),
    removeExpenses: () => dispatch(removeExpenses({id: props.expense.id}))
  }
}

export default connect(mapStateToProp, mapDispatchToProps)(EditExpensePage);
