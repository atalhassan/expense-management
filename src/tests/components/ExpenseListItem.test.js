import React from 'react'
import {shallow} from 'enzyme'
import {ExpenseListItem} from '../../components/ExpenseListItem'
import expenses from '../dummyData/expenses'

test('should render ExpenseListItem with data', () => {
  const wrapper = shallow(<ExpenseListItem {...expenses[0]}/>)
  expect(wrapper).toMatchSnapshot()
})
