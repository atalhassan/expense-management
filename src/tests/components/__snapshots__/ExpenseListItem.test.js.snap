// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem with data 1`] = `
<Link
  className="list-item "
  replace={false}
  to="/edit/1"
>
  <div>
    <h3
      className="list-item__title"
    >
      Test 1
    </h3>
    <span
      className="list-item__subtitle"
    >
      Jan 1st, 1970
    </span>
  </div>
  <h3
    className="list-item__data"
  >
    $10.00
  </h3>
</Link>
`;
